File: Machine Learning using Python and R/Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression.py
# Data Preprocessing

# 1. Importing the libraries
import numpy as np # Contains mathematical tools
import matplotlib.pyplot as plt # Contains plotting tools
import pandas as pd # Contains dataset import/manage tools

# 2. Importing the dataset
dataset = pd.read_csv('Salary_Data.csv')

# 3. Divide the dataset into independent(X) and dependent(y) variables
X = dataset.iloc[:, :-1].values # Matrix of features/variables
y = dataset.iloc[:, 1].values # Matrix of outputs/prediction/result

# Data Manipulation and Cleaning
# 4. Take care of missing data
from sklearn.preprocessing import Imputer # Imputer is a class that takes care of missing data
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer = imputer.fit(X[:, 0:1])
X[:, 0:1] = imputer.transform(X[:, 0:1])

# 5. Encode categorical data
# Encoding the Independent Variable
"""from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])
onehotencoder = OneHotEncoder(categorical_features = [0])
X = onehotencoder.fit_transform(X).toarray()
# Encoding the Dependent Variable
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)"""

# 6. Feature Scaling (Standardisation (StandardScaler) or Normalisation (Normalizer))
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X = sc_X.fit_transform(X)
sc_y = StandardScaler()
y = sc_y.fit_transform(y)"""

# Test and Train Samples
# 7. Splitting the dataset into the Training set and Test set
# from sklearn.cross_validation import train_test_split # cross_validation is deprecated and moved to model_selection
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

# Model Development - Simple Regression
# 8. Fitting Simple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# 9. Predicting the Test set results
y_pred = regressor.predict(X_test)

# 10. Evaluate/Diagnose the Model - Mean Absolute Percentage Error (MAPE)

# 11. Model Visualization
# Visualising the Training set results
plt.scatter(X_train, y_train, color = 'green')
plt.scatter(X_test, y_test, color = 'red')
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
plt.title('Salary vs Experience (Green - Training set, Red - Test set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()

File: Machine Learning using Python and R/Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression.R
# Data Preprocessing
# 1. Import libraries
# - Not needed in R

# 2. Importing the dataset
dataset = read.csv('Salary_Data.csv')

# 3. Split data into features/independent variables(X) and dependent(y) variables
# - Not needed in R

# Data Manipulation and Cleaning
# 4. Taking care of missing data
dataset$YearsExperience = ifelse(is.na(dataset$YearsExperience),
                                 ave(dataset$YearsExperience, FUN = function(x) mean(x, na.rm = TRUE)),
                                 dataset$YearsExperience)
dataset$Salary = ifelse(is.na(dataset$Salary),
                        ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                        dataset$Salary)

# # 5. Encoding categorical data
# dataset$Country = factor(dataset$Country,
#                          levels = c('France', 'Spain', 'Germany'),
#                          labels = c(1, 2, 3))
# dataset$Purchased = factor(dataset$Purchased,
#                            levels = c('Yes', 'No'),
#                            labels = c(1, 0))

# # 6. Feature Scaling - helps in rapid conversion of ML Models
# dataset[, 2:3] = scale(dataset[, 2:3])
# dataset[, 2:3] = scale(dataset[, 2:3])

# Train and test samples
# 7. Splitting the dataset into the Training set and Test set
# install.packages('caTools') - Install package caTools from packages tab or use command
# library(caTools) - Selects and imports the library
library(caTools)
set.seed(123) # 123 can be any random number
split = sample.split(dataset$Salary, SplitRatio = 2/3) # Split using dependent variable (in this case, Salary)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)
# Ensure size(Training_set) > size(Test_set)

# Model Development - Simple Regression
# 8. Fitting Simple Linear Regression to the Training set
regressor = lm(formula = Salary ~ YearsExperience,
               data = training_set)

# 9. Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)

# 10. Evaluate/Diagnose the Model - Mean Absolute Percentage Error (MAPE)
# Look at the information in Regressor
# summary(regressor) 
# - Look for Signif. codes (more starts --> better)
# - Look for P value (<0.005 is good)

# 11. Model Visualization
# Visualising the Training set results
# install.packages('ggplot2') - Install package ggplot2 from packages tab or use command
# library(ggplot2) - Selects and imports the library
library(ggplot2)
ggplot() +
  geom_point(aes(x = training_set$YearsExperience, y = training_set$Salary),
             colour = 'green') +
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary),
             colour = 'red') +
  ggtitle('Salary vs Experience(Green - Training Set, Red - Test Set)') +
  xlab('Years of experience') +
  ylab('Salary')
  